File: readability/readability.c
# include <stdio.h>
# include <cs50.h>
# include <string.h>
# include <ctype.h>
# include <math.h>

int count_letters(string s);
int count_words(string s);
int count_sentences(string s);
int calculate_index(int l, int w, int s);

int main(void)
{
    // calculate the readability index of one or more sentences.
    int grade = 0;
    string text = get_string("Text: ");
    int letters = 0;
    int words = 0;
    int sentences = 0;
    int index = 0;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        letters = count_letters(text);
        words = count_words(text);
        sentences = count_sentences(text);
    }
    index = calculate_index(letters, words, sentences);
    if (index > 16)
    {
        printf("Grade 16+");
    }
    else if (index < 1)
    {
        printf("Before Grade 1");
    }
    else
    {
        printf("Grade %i", index);
    }
    printf("\n");
}

int count_letters(string s)
{
    // count letters in string. spaces and punctuation should be left out.
    int count = 0;
    for (int i = 0, n = strlen(s); i < n; i++)
    {
        if (isalnum(s[i]))
        {
            count++;
        }
    }
    return count;
}

int count_words(string s)
{
    // count words in a sentence.
    int count = 1;
    bool flag = 0;
    for (int i = 0, n = strlen(s); i < n; i++)
    {
        if (isspace(s[i]) && flag == 0)
        {
            count++;
            flag = 1;
        }
        else
        {
            flag = 0;
        }
    }
    return count;
}

int count_sentences(string s)
{
    // count number of sentences in string
    int count = 0;
    for (int i = 0, n = strlen(s); i < n; i++)
    {
        char c = s[i];
        if (c == '.' || c == '?' || c == '!')
        {
            count++;
        }
    }
    return count;
}

int calculate_index(int l, int w, int s)
{
    // Calculate Coleman-Liau index
    float index = 0.0;
    int grade = 0;
    // gemiddeld aantal letters per 100 woorden
    float average_L = ((float) l / (float) w) * 100.0;
    // gemiddeld aantal zinnen per 100 woorden
    float average_S = ((float) s / (float) w) * 100.0;
    // formula: index = 0.0588 * L - 0.296 * S - 15.8
    index = (0.0588 * average_L) - (0.296 * average_S) - 15.8;
    grade = round(index);
    return grade;
}

File: pset6/readability/readability.py
from cs50 import get_string
import re

# Check the readability index of a certain text


def count_letters(s):
    # count letters in string. Spaces and punctuation
    # should be left out.
    count = 0
    for l in s:
        if l.isalnum():
            count = count + 1
    return count
    

def count_words(s):
    # count words in a sentence
    # splitting them by space
    a = s.split()
    return len(a)
    

def count_sentences(s):
    # count the sentences in a
    # text. Split by punctuation marks.
    a = re.split('[.?!]', s)
    return len(a) - 1
    

def calculate_index(l, w, s):
    # calculate the Coleman-Liau index
    # calculate average letters every 100 words
    average_L = (l/w) * 100
    # calculate average sentences every 100 words
    average_S = (s/w) * 100
    # apply formula
    index = (0.0588 * average_L) - (0.296 * average_S) - 15.8
    # return the index as an integer
    return round(index)
    

def main():
    # calculate the readability index of one or more
    # sentences
    text = get_string("Text: ")
    # do the countings
    letters = count_letters(text)
    words = count_words(text)
    sentences = count_sentences(text)
    print(f'Your text has {letters} letters, {words} words and {sentences} sentences.')
    # calculate the index
    index = calculate_index(letters, words, sentences)
    # define the grades
    if index > 16:
        print("Grade 16+")
    elif index < 1:
        print("Before Grade 1")
    else:
        print(f'Grade {index}')


main()